### DISCUSSION TITLE
How do you access contexts from views sent through channels?

### DISCUSSION THREAD

**Discussion by aiyoux** [leptos-rs/leptos#3400] [Q&A] [answered]
The new leptos 0.7 release notes recommends using channels to send views rather than storing views in signals which was previously possible in 0.6. However, I'm running into problem where the context is unavailable to views sent through a channel. Here is a minimal example of what I'm trying to accomplish:

```
use leptos::prelude::*;

#[derive(Clone)]
struct MyContext(String);

#[component]
fn CanAccessContext() -> impl IntoView {
    let context = use_context::<MyContext>().expect("MyContext to be accessable");
    view! {
        <div>{format!("Initial run - Context message: {:?}", context.0)}</div>
    }
}

#[component]
fn CannotAccessContext() -> impl IntoView {
    let context = use_context::<MyContext>().expect("MyContext to be accessable");
    view! {
        <div>{format!("Subsequent - Context message: {:?}", context.0)}</div>
    }
}
#[component]
fn SendsMessage(
    tx: std::sync::mpsc::Sender<AnyView>, 
    trigger: ArcTrigger
) -> impl IntoView {
  
    let _ = tx.send(
        view! {
            <CanAccessContext/>
        }.into_any()
    );
    trigger.notify();
    
    
    let second_message = move |_| { 
        let _ = tx.send(
            view! {
                <CannotAccessContext/>
            }.into_any()
        );
        trigger.notify(); 
    };

    view! {
        <div on:click=second_message>
            This element sends a view - click to send another one
        </div>
    }
}

#[component]
fn App() -> impl IntoView {
    provide_context(MyContext(String::from("Hello from context!")));
    let trigger = ArcTrigger::new();
    let (tx, rx) = std::sync::mpsc::channel();

    view! {
        <div>
            <SendsMessage tx={tx.clone()} trigger={trigger.clone()}/>
            {move || {
                trigger.track();
                rx.try_recv().unwrap_or_else(|_| view! {
                    <div>Nothing received</div>
                }.into_any())
            }}
        </div>
    }
}

fn main() {
    console_error_panic_hook::set_once();
    mount_to_body(|| {
        view! {
            <App/>
        }
    })
}
```
The first view that is received on the channel is able to access the context, however any subsequent views received are unable to access the context. 

I've considered the alternative of manually sending props to any element sent through a channel, however my application depends a lot on the context api, this would result in significant restructuring...

Am I missing something obvious? Any help would be greatly appreciated!

**Comment by gbj** [comment, marked as answer]
The problem here is mostly just that context isn't available in event handlers, I think.

The short answer: Use `Owner::current()` to get the current owner at some point (i.e., in the body of the `SendsMessage` component, and use `owner.with(|| { ... })` or `OwnedView::new_with_owner` to set it as the owner for the view you're sending through the channel.

**Reply by aiyoux** [reply]
Great! I can confirm this has resolved my issue, much appreciated!
